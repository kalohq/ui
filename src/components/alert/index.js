/* @flow */
export {default} from './alert.react';
export {default as AlertPopover} from './components/alert-popover.react';
