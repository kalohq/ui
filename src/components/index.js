/**
 * Typography
 */
export {default as H1} from './h1';
export {default as H2} from './h2';
export {default as H3} from './h3';
export {default as H4} from './h4';
export {default as Heading} from './heading';
export {default as Text} from './text';

/**
 * Layout
 */
export * from './skeleton';
export {Box, Flex, Block, Inline, InlineFlex, A} from './layout';
export {default as Paper, PaperToolbar} from './paper';
export {default as PaperMenu, PaperMenuItem} from './paper-menu';

/**
 * Form
 */
export {default as Button} from './button';
export {default as ButtonDropdown} from './button-dropdown';
export {default as ButtonGroup} from './button-group';
export {default as Checkbox} from './checkbox';
export {default as Radio} from './radio';

/**
 * Other
 */
export {default as Alert, AlertPopover} from './alert';
export {default as Icon} from './icon';
export {default as IconSymbols} from './icon-symbols';
export {default as LoadingSpinner} from './loading-spinner';
export {default as Sticky} from './sticky';

export {default as theme} from './theme';
