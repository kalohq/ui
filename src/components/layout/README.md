# Layout

The layout components are the primitive building blocks of the Kalo User Interface.

## Box and Flex
A wrapper component that implements some default flex values.

## Inline
An `inline-block` element.

## Spacing
Margin and padding can be set via the `margin` and `padding` props. The re are also more granular props for specifying particular directions: `marginLeft`, `paddingTop` etc.

These props can take a variety of values:
- Value specific values: `padding="20px"`
- Or scale values: `margin="medium"`

The scale values relate to a multiple of 8px:

```js
marginLeft="extra-small" // margin-left: 2px
marginLeft="small" // margin-left: 4px
marginLeft="medium" // margin-left: 8px
marginLeft="large" // margin-left: 16px
marginLeft="extra-large" // margin-left: 24px
```

These can also be mixed with pixel specific values:

```js
padding={[20, "large", "medium", 38]}
```

## The css and style props
The `css` prop is inherited from [Emotion](https://github.com/emotion-js/emotion) and is available on every component. Styles passed in through this prop, will be set via a classname generated at run-time.

The `style` prop is used to add styles to an element inline via the style attribute. This should be used for regulary computed styles as it is considerably more performant.
