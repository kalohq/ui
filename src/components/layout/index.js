/* @flow */
export {Box, Flex, Block, Inline, InlineFlex, A} from './layout';
