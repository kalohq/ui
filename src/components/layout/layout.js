/* @flow */
import React from 'react';
import {
  parseStyleProps,
  cleanProps,
  spacing,
  filterStyleProps,
} from '../../utils/style';
import PureComponent from 'react-pure-render/component';
import styled, {cx} from 'react-emotion';

const makePrimitive = (name, DefaultComponent) => {
  class Primitive extends PureComponent {
    render() {
      const {
        component: Component = DefaultComponent,
        elRef,
        style: propStyle,
        className,
        ...otherProps
      } = this.props;

      const {props, style} = parseStyleProps({
        ...otherProps,
      });

      const cleanedProps = cleanProps(props);

      return (
        <Component
          ref={elRef}
          className={className ? cx(className) : null}
          css={{...style}}
          {...cleanedProps}
          style={{...propStyle}}
        />
      );
    }
  }

  Primitive.displayName = `Primitive${name}`;
  return Primitive;
};

/** 
 * Layout primitives
 */

export const Box = styled(makePrimitive('Box', 'div'))`
  position: relative;
  flex-direction: column;
  align-items: stretch;
  flex-shrink: 0;
  align-content: flex-start;
  display: flex;
  ${spacing};
  ${filterStyleProps};
`;
Box.displayName = 'Box';

export const Flex = styled(makePrimitive('Flex', 'div'))`
  display: flex;
  ${spacing};
  ${filterStyleProps};
`;
Flex.displayName = 'Flex';

export const Block = styled(makePrimitive('Block', 'div'))`
  display: block;
  ${spacing};
  ${filterStyleProps};
`;
Block.displayName = 'Block';

export const Inline = styled(makePrimitive('Inline', 'span'))`
  display: inline-block;
  vertical-align: bottom;
  ${spacing};
  ${filterStyleProps};
`;
Inline.displayName = 'Inline';

export const InlineFlex = styled(makePrimitive('InlineFlex', 'span'))`
  display: inline-flex;
  ${spacing};
  ${filterStyleProps};
`;
InlineFlex.displayName = 'InlineFlex';

/**
 * All of our html primitives should follow here so our api is consistent to the very root
 * Eg. Button, Input, Select, Table etc.
 */

export const A = styled(makePrimitive('A', 'a'))`
  display: inline-flex;
  ${spacing};
  ${filterStyleProps};
`;
A.displayName = 'A';
