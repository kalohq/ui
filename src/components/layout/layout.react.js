import styled from 'react-emotion';
import {makePrimitive} from './utils';

/**
 * Layout primitives
 */

export const Box = styled(
  makePrimitive('Box', 'div', {
    position: 'relative',
    flexDirection: 'column',
    alignItems: 'stretch',
    flexShrink: 0,
    alignContent: 'flex-start',
    display: 'flex',
  })
)();
Box.displayName = 'Box';

export const Flex = styled(makePrimitive('Flex', 'div', {display: 'flex'}))();
Flex.displayName = 'Flex';

export const Block = styled(
  makePrimitive('Block', 'div', {display: 'block'})
)();
Block.displayName = 'Block';

export const Inline = styled(
  makePrimitive('Inline', 'span', {
    display: 'inline-block',
    verticalAlign: 'bottom',
  })
)();
Inline.displayName = 'Inline';

export const InlineFlex = styled(
  makePrimitive('InlineFlex', 'span', {display: 'inline-flex'})
)();
InlineFlex.displayName = 'InlineFlex';

/**
 * All of our html primitives should follow here so our api is consistent to the very root
 * Eg. Button, Input, Select, Table etc.
 */

export const A = styled(makePrimitive('A', 'a', {display: 'inline-flex'}))();
A.displayName = 'A';

/**
 * This is an empty component that provides backwards compatibility as we phase out the
 * layout primivites above. Its sole purpose is is to convert inline styles props in to
 * actual CSS.
 */

export const UIBase = styled(makePrimitive('UIBase', 'div'))();
UIBase.displayName = 'UIBase';
