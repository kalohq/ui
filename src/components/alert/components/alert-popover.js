/* @flow */
import * as React from 'react';
import {Box} from '../../layout';
import CSSTransitionGroup from 'react-addons-css-transition-group';

type TProps = {
  children: React.Element<*>,
  topOffset?: number,
  zIndex?: number,
};

export default function AlertPopover(props: TProps) {
  const {children, topOffset = 0, zIndex = 999} = props;

  return (
    <CSSTransitionGroup
      transitionName="t-slide-from-right"
      transitionAppear={true}
      transitionAppearTimeout={300}
      transitionEnterTimeout={300}
      transitionLeaveTimeout={300}
      component={Box}
      position="fixed"
      right={32}
      top={topOffset + 16}
      maxWidth={400}
      alignItems="flex-end"
      zIndex={zIndex}
    >
      {React.Children.map(
        children,
        child =>
          child ? (
            <Box marginBottom={8}>
              {React.cloneElement(child, {justifyContent: 'left'})}
            </Box>
          ) : null
      )}
    </CSSTransitionGroup>
  );
}
