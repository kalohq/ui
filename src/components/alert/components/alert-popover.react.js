import * as React from 'react';
import PropTypes from 'prop-types';
import CSSTransitionGroup from 'react-addons-css-transition-group';

import {Box} from '../../layout';

const AlertPopover = props => {
  const {children, topOffset = 0, zIndex = 999} = props;
  return (
    <CSSTransitionGroup
      transitionName="t-slide-from-right"
      transitionAppear={true}
      transitionAppearTimeout={300}
      transitionEnterTimeout={300}
      transitionLeaveTimeout={300}
      component={Box}
      position="fixed"
      right={32}
      top={topOffset + 16}
      maxWidth={400}
      alignItems="flex-end"
      zIndex={zIndex}
    >
      {React.Children.map(children, child =>
        child ? (
          <Box marginBottom={8}>
            {React.cloneElement(child, {justifyContent: 'left'})}
          </Box>
        ) : null
      )}
    </CSSTransitionGroup>
  );
};

AlertPopover.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.arrayOf(PropTypes.node).isRequired,
  ]),
  topOffset: PropTypes.number,
  zIndex: PropTypes.number,
};

export default AlertPopover;
