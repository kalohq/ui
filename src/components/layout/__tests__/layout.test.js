/* @flow */
/* eslint-env jest */
import * as React from 'react';
import * as emotion from 'emotion';
import {createSerializer} from 'jest-emotion';
import renderer from 'react-test-renderer';
import {Box, Block, Flex, Inline, InlineFlex} from 'components/layout';

expect.addSnapshotSerializer(createSerializer(emotion));

describe('Box', () => {
  const create = (props = {}) => renderer.create(<Box {...props} />).toJSON();

  test('should render correctly', () => {
    const element = create();
    expect(element).toMatchSnapshot();
  });

  test('should allow overriding of default styles', () => {
    const element = create({
      flex: 0,
      alignContent: 'flex-end',
      flexShrink: 1,
      style: {color: 'red'},
    });

    expect(element).toMatchSnapshot();
  });

  test('should allow passing vector style values', () => {
    const element = create({
      margin: [0, 5, '10%', 15],
      padding: [15, undefined],
    });
    expect(element).toMatchSnapshot();
  });

  test('should allow passing spacing scales', () => {
    const element = create({
      margin: ['small', 'medium', 'large'],
      paddingTop: 'small',
    });

    expect(element).toMatchSnapshot();
  });
});

describe('Block', () => {
  const create = (props = {}) => renderer.create(<Block {...props} />).toJSON();

  test('should render correctly', () => {
    const element = create({
      flex: 0,
      onClick: null,
    });
    expect(element).toMatchSnapshot();
  });
});

describe('Flex', () => {
  const create = (props = {}) => renderer.create(<Flex {...props} />).toJSON();

  test('should render correctly', () => {
    const element = create({
      flex: 0,
      onClick: null,
    });
    expect(element).toMatchSnapshot();
  });
});

describe('Inline', () => {
  const create = (props = {}) =>
    renderer.create(<Inline {...props} />).toJSON();

  test('should render correctly', () => {
    const element = create({
      flex: 0,
      onClick: null,
    });
    expect(element).toMatchSnapshot();
  });
});

describe('InlineFlex', () => {
  const create = (props = {}) =>
    renderer.create(<InlineFlex {...props} />).toJSON();

  test('should render correctly', () => {
    const element = create({
      flex: 0,
      onClick: null,
    });
    expect(element).toMatchSnapshot();
  });
});
