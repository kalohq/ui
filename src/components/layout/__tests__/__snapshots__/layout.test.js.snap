// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Block should render correctly 1`] = `
.glamor-1 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: block;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
}

<div
  className="glamor-0 glamor-1"
  onClick={null}
  style={Object {}}
/>
`;

exports[`Box should allow overriding of default styles 1`] = `
.glamor-1 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  -webkit-align-content: flex-end;
  -ms-flex-line-pack: flex-end;
  align-content: flex-end;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: flex-start;
  align-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  -webkit-align-content: flex-end;
  -ms-flex-line-pack: flex-end;
  align-content: flex-end;
  -webkit-flex-shrink: 1;
  -ms-flex-negative: 1;
  flex-shrink: 1;
}

<div
  className="glamor-0 glamor-1"
  style={
    Object {
      "color": "red",
    }
  }
/>
`;

exports[`Box should allow passing spacing scales 1`] = `
.glamor-1 {
  margin: 4px 8px 16px;
  padding-top: 4px;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: flex-start;
  align-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 4px 8px 16px;
  padding-top: 4px;
}

<div
  className="glamor-0 glamor-1"
  style={Object {}}
/>
`;

exports[`Box should allow passing vector style values 1`] = `
.glamor-1 {
  margin: 0px 5px 10% 15px;
  padding: 15px;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: flex-start;
  align-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  margin: 0px 5px 10% 15px;
  padding: 15px;
}

<div
  className="glamor-0 glamor-1"
  style={Object {}}
/>
`;

exports[`Box should render correctly 1`] = `
.glamor-1 {
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-align-content: flex-start;
  -ms-flex-line-pack: flex-start;
  align-content: flex-start;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div
  className="glamor-0 glamor-1"
  style={Object {}}
/>
`;

exports[`Flex should render correctly 1`] = `
.glamor-1 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
}

<div
  className="glamor-0 glamor-1"
  onClick={null}
  style={Object {}}
/>
`;

exports[`Inline should render correctly 1`] = `
.glamor-1 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: inline-block;
  vertical-align: bottom;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
}

<span
  className="glamor-0 glamor-1"
  onClick={null}
  style={Object {}}
/>
`;

exports[`InlineFlex should render correctly 1`] = `
.glamor-1 {
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-flex: 0;
  -ms-flex: 0;
  flex: 0;
}

<span
  className="glamor-0 glamor-1"
  onClick={null}
  style={Object {}}
/>
`;
